libevmasm/ConstantOptimiser.cpp

```diff
@@ -144,30 +144,47 @@ bigint CodeCopyMethod::gasNeeded() const
 		// Data gas for copy routines: Some bytes are zero, but we ignore them.
 		bytesRequired(copyRoutine()) * (m_params.isCreation ? GasCosts::txDataNonZeroGas : GasCosts::createDataGas),
 		// Data gas for data itself
-		dataGas(toBigEndian(m_value))
+		dataGas(toCompactBigEndian(m_value, 1))
 	);
 }
 
 AssemblyItems CodeCopyMethod::execute(Assembly& _assembly) const
 {
-	bytes data = toBigEndian(m_value);
+	bytes data = toCompactBigEndian(m_value, 1);
+	assertThrow(data.size() >= 1 && data.size() <= 32, OptimizerException, "Invalid number encoding.");
 	AssemblyItems actualCopyRoutine = copyRoutine();
 	actualCopyRoutine[4] = _assembly.newData(data);
+	if (data.size() != 32)
+	{
+		// Update sizes for short encodings.
+		actualCopyRoutine[3] = u256(data.size());
+		actualCopyRoutine[5] = u256(32 - data.size());
+	}
 	return actualCopyRoutine;
 }
 
 AssemblyItems const& CodeCopyMethod::copyRoutine()
 {
 	AssemblyItems static copyRoutine{
+		// constant to be reused 3+ times
 		u256(0),
+
+		// back up memory
+		// mload(0)
 		Instruction::DUP1,
-		Instruction::MLOAD, // back up memory
-		u256(32),
-		AssemblyItem(PushData, u256(1) << 16), // has to be replaced
-		Instruction::DUP4,
+		Instruction::MLOAD,
+
+		// codecopy(0, <pushdata>, 32)
+		u256(32), // replaced above in actualCopyRoutine[3]
+		AssemblyItem(PushData, u256(1) << 16), // replaced above in actualCopyRoutine[4]
+		Instruction::DUP4, // replaced above in actualCopyRoutine[5]
```

Comment:
For 100% correctness in gas estimation in `gasNeeded` this should be `PUSH32`, however I'd say the cost here is negligible and would require another conditional in `execute`. If we really want to be correct we'd need to split `CodeCopyMethod` into two: a compact and a non-compact version. While it is not a lot of code to do it, not sure about the returns this brings us.